### DISCUSSION TITLE
ERROR (payload): Forbidden: You are not allowed to perform this action.

### DISCUSSION THREAD

**Discussion by eduhdev12** [payloadcms/payload#1918] [Q&A] [answered]
# Bug Report

Hi, I deployed my payload-cms to docker on a vps, but after the first user registration I don't have access to anything. Everything works fine, no other errors in the container logs, but I can't create any new data in the collection due to `"ERROR (payload): Forbidden: You are not allowed to perform this action".`
This only happens in a production environment, but running `yarn serve` on a local machine doesn't present any problems.

## Steps to Reproduce

1. Account registration process
2. Creation of a new category attempt

## Other Details

Payload version: 1.5.9

**Comment by jmikrut** [comment]
Hey @eduhdev12 — this is probably related to one of two things:

1. your cookie config. You should open your inspector, look at the Network tab, and inspect the auth cookie that is being set. Make sure it doesn't get blocked for any reason. It could be that you have some cookie settings that are conflicting with your live instance. 
2. CSRF. Take a look at this link: https://payloadcms.com/docs/authentication/overview#csrf-protection - you may need to whitelist your production domain within your `csrf` settings to allow the cookie to be read while on your production domain.

I'll convert this to a discussion to keep conversation going as this is not a problem with Payload itself, but I'm more than happy to help!

**Reply by eduhdev12** [reply]
Thank you for these details, but sadly the `csrf` setting didn't fix the issue and I can confirm that `cookie` header is present in request!

```
cookie: payload-token=jwt
```
The decoded jwt payload is good, but still no new results!
I really want to fix this issue, it's very important

**Reply by jmikrut** [reply]
Are there any yellow "alert" symbols on the cookie? what browser are you using? 

You should compare the Payload demo network panel cookies / requests to that of your prod instance.

See my inspector here after logging into the demo:

<img width="1728" alt="Screen Shot 2023-01-19 at 11 25 51 AM" src="https://user-images.githubusercontent.com/6651896/213497665-1f7b36c8-3a03-441b-9b01-7c2a15e56d59.png">

**Reply by eduhdev12** [reply]
![mdJVFubdSr](https://user-images.githubusercontent.com/60825949/213499021-1e5e673a-11b8-4654-beea-22cc772a7387.png)

The `/api/access` request has no issues, the problem is when I try to create a new item in any collection even that create access is `() => true`

**Reply by eduhdev12** [reply]
This is the whole console log if it helps
```
[16:28:49] ERROR (payload): Forbidden: You are not allowed to perform this action.
    at new ExtendableError (/usr/src/app/node_modules/payload/dist/errors/APIError.js:22:15)
    at new APIError (/usr/src/app/node_modules/payload/dist/errors/APIError.js:38:9)
    at new Forbidden (/usr/src/app/node_modules/payload/dist/errors/Forbidden.js:10:9)
    at executeAccess (/usr/src/app/node_modules/payload/dist/auth/executeAccess.js:9:23)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async create (/usr/src/app/node_modules/payload/dist/collections/operations/create.js:37:9)
    at async createHandler (/usr/src/app/node_modules/payload/dist/collections/requestHandlers/create.js:12:21)
```

**Reply by thenaazimkhan** [reply]
I am also facing the same issue. for me, all the access (create, update, delete) is working fine on my local dev but when I push my app to production the access (create, update, delete) control is through the "Forbidden: You are not allowed to perform this action." read is working fine on prod.

https://github.com/payloadcms/payload/discussions/2722

**Reply by Barbacoa08** [reply]
@thenaazimkhan, my issue was resolved by adding `http`+`https` to my `cors` and `csrf` rules in my payload config (example image attached), as described in [the selected answer](https://github.com/payloadcms/payload/discussions/1918#discussioncomment-4781784) for this discussion.

![payload config example](https://github.com/payloadcms/payload/assets/13157284/d0a7efad-d64a-465a-959c-2fd4aa319c8d)

**Reply by thenaazimkhan** [reply]
@Barbacoa08, thanks for your reply.
in your config, you are using User collection for the admin login. but in my case, I have two collections, one is Admin only for the admin user and one is User for the frontend user login.

My issue is resolved after removing the serverURL from payload.config.ts, access control is working fine except for Admin collection means all the access is applying to the other collection but when I am applying access control on my Admin collection, the access is not working.
note: all working fine in dev env only issue is on prod.

payload.config.ts
<img width="654" alt="image" src="https://github.com/payloadcms/payload/assets/95960950/5c72c3b5-dba5-4713-befb-a9a34fe620b5">

collection/index.ts
<img width="378" alt="image" src="https://github.com/payloadcms/payload/assets/95960950/fa51eb1f-e641-4c6b-a9b8-0a389b7ea242">

Admins.ts
<img width="309" alt="image" src="https://github.com/payloadcms/payload/assets/95960950/3c7f2fa4-c53a-4378-b4a5-fb0364d9c4b8">

this is for login into the admin panel. you can see I have put console.log in access.read fn. I can see the log in dev env but when i pushed the same to my production I can not see any log means.

Users.ts
<img width="594" alt="image" src="https://github.com/payloadcms/payload/assets/95960950/2ae91293-7ddd-4482-81f6-44a2f391d38f">

frontend user login

**Reply by Barbacoa08** [reply]
Oof, sorry I can't help with that. Your setup is much more complicated than mine and I haven't had to any of that, sorry I can't help more 😞

**Reply by thenaazimkhan** [reply]
np thanks, man...

**Comment by Barbacoa08** [comment]
This sounds like the same issue that I'm running into ([link to my discussion](https://github.com/payloadcms/payload/discussions/1904)).

Related, the [deployment docs](https://payloadcms.com/docs/deployment#Docker) is 404.

**Reply by eduhdev12** [reply]
Have you managed to fix it?

**Reply by Barbacoa08** [reply]
Nope 😞. Still poking around trying to get it working. I can use `() => true` to be able to make updates. So for now I'm just setting everything to `() => true` while I make updates, than immediately reseting it back to `isAdmin`/ect once I'm done making edits. Or just editing locally, as I'm also _only_ seeing this happen in the prod environment, localhost works as expected.

**Reply by eduhdev12** [reply]
For me `() => true` is not working at all in production, just on local env

**Reply by Barbacoa08** [reply]
Your `payload.config.ts` has this line in it yes?
`serverURL: process.env.PAYLOAD_PUBLIC_BASE_DNS`

And your `PAYLOAD_PUBLIC_BASE_DNS` includes your UN+PW yes? MongoDB Atlas will look like this:
`mongodb+srv://{username}&{password}@{dbname}.ntzanfx.mongodb.net/?retryWrites=true&w=majority`

BTW, the only tutorial I could find about hosting PayloadCMS was [at Northflank](https://northflank.com/guides/deploying-payload-cms). I'm running into a similar issue that you are, but I was able to get a bit further. And, it's free for something as simple as a POC.

**Reply by eduhdev12** [reply]
I am trying to see if the serverURL is the issue, but I am sure that mongodb is working because I tried to change something manually in database and it's reading the date, the problem is writing new data

Do you include in the serverURL the `http`/`https` method?

**Reply by Barbacoa08** [reply]
`https`

**Reply by eduhdev12** [reply]
I made a little progress, now `() => true` works good, but any access guard or checking if `user` is undefined is not working at all, doing requests from the cms and I get the user as undefined all the time

**Reply by Barbacoa08** [reply]
Are you using a docker container? It sounds like you're in basically the same situation that I am now, and I'm wondering how close our projects are. 

My hosting environment is running a docker container. And I'm wondering if it's the docker container that's causing these auth issues. 

Also, do you have any `csrf` and/or `cors` settings in your `payload.config.ts`?

**Reply by eduhdev12** [reply]
Yes, I am using a normal docker container and `csrf`,  `cors: "*"`, sometimes it makes a double call to create access and the first one is undefined and the second one contains the user, but the error comes up without any action.

**Reply by Barbacoa08** [reply]
Yup, that's what I'm seeing as well. I have my `csrf` and `cors` locked to my environment, but otherwise we seem to have the same setup and are seeing the same thing happen.

**Comment by eduhdev12** [comment, marked as answer]
Hello, in my case someone helped me and changed the `csrf` that must follow the prefix connection type if it's `http` or `https` and now seems to work flawlessly!
